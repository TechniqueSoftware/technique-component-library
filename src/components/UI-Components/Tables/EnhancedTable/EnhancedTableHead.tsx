import * as React from 'react';
import { createStyles, makeStyles } from '@material-ui/core/styles';
import TableCell from '@material-ui/core/TableCell';
import TableHead from '@material-ui/core/TableHead';
import TableRow from '@material-ui/core/TableRow';
import TableSortLabel from '@material-ui/core/TableSortLabel';
import { Checkbox } from '../../Checkbox/Checkbox';
import { HeaderCellProps, OrderDirection } from './EnhancedTable';

const useStyles = makeStyles(() =>
  createStyles({
    visuallyHidden: {
      border: 0,
      clip: 'rect(0 0 0 0)',
      height: 1,
      margin: -1,
      overflow: 'hidden',
      padding: 0,
      position: 'absolute',
      top: 20,
      width: 1,
    },
  }),
);

export interface EnhancedTableHeadProps {
  headerCellProps: HeaderCellProps[];
  enableSelectableRows: boolean;
  numSelected: number;
  onSelectAllClick: (event: React.ChangeEvent<HTMLInputElement>, checked: boolean) => void;
  order: OrderDirection;
  orderBy: string;
  rowCount: number;
  onRequestSort: (event: React.MouseEvent<unknown>, property: string) => void;
  rowDescriptor?: string;
}

export function EnhancedTableHead(props: EnhancedTableHeadProps) {
  const {
    enableSelectableRows,
    numSelected,
    rowCount,
    orderBy,
    order,
    onRequestSort,
    headerCellProps,
    onSelectAllClick,
    rowDescriptor
  } = props;

  const createSortHandler = (property: string) => (event: React.MouseEvent<unknown>) => {
    onRequestSort(event, property);
  };
  const classes = useStyles();

  return (
    <TableHead>
      <TableRow>
        {enableSelectableRows && <TableCell padding="checkbox">
          <Checkbox
            data-test={'select-all-checkbox'}
            indeterminate={numSelected > 0 && numSelected < rowCount}
            checked={numSelected === rowCount}
            onChange={onSelectAllClick}
            inputProps={{ 'aria-label': `select all ${rowDescriptor}` }}
          />
        </TableCell>}
        {headerCellProps.map(header => (
          <TableCell
            size={'small'}
            key={header.id}
            align={header.align}
            sortDirection={orderBy === header.id ? order : false}
          >
            {header.sortable ?
              <TableSortLabel
                hideSortIcon={!header.sortable}
                active={orderBy === header.id}
                direction={order}
                onClick={createSortHandler(header.id)}
                data-test={`table-sort-label-${header.id}`}
              >
                {header.element}
                {orderBy === header.id ? (
                  <span className={classes.visuallyHidden}>
                  {order === 'desc' ? 'sorted descending' : 'sorted ascending'}
                </span>
                ) : null}
              </TableSortLabel>
              : header.element}
          </TableCell>
        ))}
      </TableRow>
    </TableHead>
  );
}
