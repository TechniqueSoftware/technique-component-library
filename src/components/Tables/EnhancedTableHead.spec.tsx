import { EnhancedTableHead, EnhancedTableHeadProps } from './EnhancedTableHead';
import * as React from 'react';
import { shallow } from 'enzyme';

describe('EnhancedTableHead', () => {

  const baseProps: EnhancedTableHeadProps = {
    rowDescriptor: 'mockRowDescriptor',
    onRequestSort: jest.fn(),
    onSelectAllClick: jest.fn(),
    rowCount: 10,
    numSelected: 0,
    enableSelectableRows: false,
    order: 'desc',
    orderBy: 'column1',
    headerCellProps: [
      { id: 'column1', sortable: true, align: 'right', element: <span>column1Element</span> },
      { id: 'column2', sortable: false, align: 'left', element: <span>column2Element</span> }
    ]
  };

  test('should render with minimal props', () => {
    const shallowWrapper = shallow(<EnhancedTableHead {...baseProps} />);
    expect(shallowWrapper.debug()).toMatchSnapshot();
  });

  test('should call sort handler on click and return the id', () => {
    const shallowWrapper = shallow(<EnhancedTableHead {...baseProps} />);
    const id = baseProps.headerCellProps[0].id;
    const tableSortLabel = shallowWrapper.find(`[data-test="table-sort-label-${id}"]`);
    tableSortLabel.simulate('click');
    expect((baseProps.onRequestSort as any).mock.calls).toEqual([[undefined, id]]);
    expect(shallowWrapper.debug()).toMatchSnapshot();
  });

  test('should show select all box and handle clicking it when enableSelectableRows is true', () => {
    const props: EnhancedTableHeadProps = {
      ...baseProps,
      enableSelectableRows: true,
      onSelectAllClick: jest.fn()
    };
    const shallowWrapper = shallow(<EnhancedTableHead {...props} />);
    const selectAllCheckbox = shallowWrapper.find(`[data-test="select-all-checkbox"]`);
    selectAllCheckbox.simulate('change');
    expect((props.onSelectAllClick as any).mock.calls.length).toBe(1);
    expect(shallowWrapper.debug()).toMatchSnapshot();
  });
});
