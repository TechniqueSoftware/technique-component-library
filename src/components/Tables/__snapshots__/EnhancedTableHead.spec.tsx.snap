// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EnhancedTableHead should call sort handler on click and return the id 1`] = `
"<WithStyles(ForwardRef(TableHead))>
  <WithStyles(ForwardRef(TableRow))>
    <WithStyles(ForwardRef(TableCell)) size=\\"small\\" align=\\"right\\" sortDirection=\\"desc\\">
      <WithStyles(ForwardRef(TableSortLabel)) hideSortIcon={false} active={true} direction=\\"desc\\" onClick={[Function (anonymous)]} data-test=\\"table-sort-label-column1\\">
        <span>
          column1Element
        </span>
        <span className={{...}}>
          sorted descending
        </span>
      </WithStyles(ForwardRef(TableSortLabel))>
    </WithStyles(ForwardRef(TableCell))>
    <WithStyles(ForwardRef(TableCell)) size=\\"small\\" align=\\"left\\" sortDirection={false}>
      <span>
        column2Element
      </span>
    </WithStyles(ForwardRef(TableCell))>
  </WithStyles(ForwardRef(TableRow))>
</WithStyles(ForwardRef(TableHead))>"
`;

exports[`EnhancedTableHead should render with minimal props 1`] = `
"<WithStyles(ForwardRef(TableHead))>
  <WithStyles(ForwardRef(TableRow))>
    <WithStyles(ForwardRef(TableCell)) size=\\"small\\" align=\\"right\\" sortDirection=\\"desc\\">
      <WithStyles(ForwardRef(TableSortLabel)) hideSortIcon={false} active={true} direction=\\"desc\\" onClick={[Function (anonymous)]} data-test=\\"table-sort-label-column1\\">
        <span>
          column1Element
        </span>
        <span className={{...}}>
          sorted descending
        </span>
      </WithStyles(ForwardRef(TableSortLabel))>
    </WithStyles(ForwardRef(TableCell))>
    <WithStyles(ForwardRef(TableCell)) size=\\"small\\" align=\\"left\\" sortDirection={false}>
      <span>
        column2Element
      </span>
    </WithStyles(ForwardRef(TableCell))>
  </WithStyles(ForwardRef(TableRow))>
</WithStyles(ForwardRef(TableHead))>"
`;

exports[`EnhancedTableHead should show select all box and handle clicking it when enableSelectableRows is true 1`] = `
"<WithStyles(ForwardRef(TableHead))>
  <WithStyles(ForwardRef(TableRow))>
    <WithStyles(ForwardRef(TableCell)) padding=\\"checkbox\\">
      <Checkbox data-test=\\"select-all-checkbox\\" indeterminate={false} checked={false} onChange={[Function: mockConstructor] { _isMockFunction: true, getMockImplementation: [Function (anonymous)], mock: Object [Object: null prototype] { calls: [ [] ], instances: [ undefined ], invocationCallOrder: [ 2 ], results: [ Object [Object: null prototype] { isThrow: false, value: undefined } ] }, mockClear: [Function (anonymous)], mockReset: [Function (anonymous)], mockRestore: [Function (anonymous)], mockReturnValueOnce: [Function (anonymous)], mockResolvedValueOnce: [Function (anonymous)], mockRejectedValueOnce: [Function (anonymous)], mockReturnValue: [Function (anonymous)], mockResolvedValue: [Function (anonymous)], mockRejectedValue: [Function (anonymous)], mockImplementationOnce: [Function (anonymous)], mockImplementation: [Function (anonymous)], mockReturnThis: [Function (anonymous)], mockName: [Function (anonymous)], getMockName: [Function (anonymous)] }} inputProps={{...}} />
    </WithStyles(ForwardRef(TableCell))>
    <WithStyles(ForwardRef(TableCell)) size=\\"small\\" align=\\"right\\" sortDirection=\\"desc\\">
      <WithStyles(ForwardRef(TableSortLabel)) hideSortIcon={false} active={true} direction=\\"desc\\" onClick={[Function (anonymous)]} data-test=\\"table-sort-label-column1\\">
        <span>
          column1Element
        </span>
        <span className={{...}}>
          sorted descending
        </span>
      </WithStyles(ForwardRef(TableSortLabel))>
    </WithStyles(ForwardRef(TableCell))>
    <WithStyles(ForwardRef(TableCell)) size=\\"small\\" align=\\"left\\" sortDirection={false}>
      <span>
        column2Element
      </span>
    </WithStyles(ForwardRef(TableCell))>
  </WithStyles(ForwardRef(TableRow))>
</WithStyles(ForwardRef(TableHead))>"
`;
